river/Seat.zig

```diff
@@ -342,28 +342,53 @@ pub fn handleViewUnmap(self: *Self, view: *View) void {
     if (self.focused == .view and self.focused.view == view) self.focus(null);
 }
 
-/// Handle any user-defined mapping for the passed keysym and modifiers
+/// Handle any user-defined mapping for the passed keysym/keycode and modifiers
 /// Returns true if the key was handled
 pub fn handleMapping(
     self: *Self,
-    keysym: xkb.Keysym,
-    modifiers: wlr.Keyboard.ModifierMask,
+    keycode: xkb.Keycode,
+    keysyms_raw: []const xkb.Keysym,
+    modifiers_raw: wlr.Keyboard.ModifierMask,
+    keysyms_translated: []const xkb.Keysym,
+    modifiers_translated: wlr.Keyboard.ModifierMask,
     released: bool,
 ) bool {
     const modes = &server.config.modes;
-    for (modes.items[self.mode_id].mappings.items) |*mapping| {
-        if (std.meta.eql(modifiers, mapping.modifiers) and keysym == mapping.keysym and released == mapping.release) {
-            if (mapping.repeat) {
-                self.repeating_mapping = mapping;
-                self.mapping_repeat_timer.timerUpdate(server.config.repeat_delay) catch {
-                    log.err("failed to update mapping repeat timer", .{});
-                };
+
+    const mapping_match: ?*Mapping = blk: {
+        // `to-code` mappings take precedence
+        for (modes.items[self.mode_id].mappings.items) |*mapping| {
+            if (mapping.matchCode(keycode, modifiers_raw, released)) {
+                break :blk mapping;
             }
-            self.runCommand(mapping.command_args);
-            return true;
         }
-    }
-    return false;
+
+        // xkb-translated mappings take precedence over raw mappings
+        for (modes.items[self.mode_id].mappings.items) |*mapping| {
+            if (mapping.matchSym(keysyms_translated, modifiers_translated, released)) {
+                break :blk mapping;
+            }
+        }
+
+        // raw mappings
+        for (modes.items[self.mode_id].mappings.items) |*mapping| {
+            if (mapping.matchSym(keysyms_raw, modifiers_raw, released)) {
+                break :blk mapping;
+            }
+        }
+        break :blk null;
+    };
+
+    if (mapping_match) |mapping| {
+        if (mapping.repeat) {
+            self.repeating_mapping = mapping;
+            self.mapping_repeat_timer.timerUpdate(server.config.repeat_delay) catch {
+                log.err("failed to update mapping repeat timer", .{});
+            };
+        }
+        self.runCommand(mapping.command_args);
+        return true;
+    } else return false;
```

Comment:
Since the other branch of the if has braces I'd rather this one have them too to keep things visually balanced.
```suggestion
    } else {
        return false;
    }
```